javascript/ql/src/Security/CWE-116/BadTagFilter.qhelp

```diff
@@ -0,0 +1,55 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+
+<overview>
+<p>
+Parsing general HTML using regular expressions is impossible, however it is possible to match
+single HTML tags. However, if the regexp is not written well it might be easy 
+to circumvent the regexp, which can lead to XSS or other security issues.
+</p>
+<p>
+Many of these mistakes are caused by browsers having very forgiving HTML parsers: 
+Browsers will often render invalid HTML with parser errors. 
+Regular expressions that attempt to match HTML must recognize tags containing these parser errors.
+</p>
+</overview>
+
+<recommendation>
+<p>
+Use a well-tested sanitization or parser library if at all possible. These libraries are much more
+likely to handle corner cases correctly than a custom implementation.
+</p>
+</recommendation>
+
+<example>
+<p>
+For example, assume we want to write a function that filters out all <code>&lt;script&gt;</code> tags.
+Such a function might be written like below: 
+</p>
+
+<sample src="examples/BadTagFilter.js" />
+
+<p>
+This sanitizer does not filter out all <code>&lt;script&gt;</code> tags. 
+Browsers will not only accept <code>&lt;/script&gt;</code> as script end tags, but also tags such as <code>&lt;/script foo="bar"&gt;</code> even though it is a parser error.
+This means that an attack string such as <code>&lt;script&gt;alert(1)&lt;/script foo="bar"&gt;</code> will not be filtered by 
+the function, but <code>alert(1)</code> will be executed by a browser if the string is rendered as HTML.
```

Comment:
```suggestion
This sanitizer does not filter out all <code>&lt;script&gt;</code> tags. 
Browsers will not only accept <code>&lt;/script&gt;</code> as script end tags, but also tags such as <code>&lt;/script foo="bar"&gt;</code> even though it is a syntax error.
This means that an attack string such as <code>&lt;script&gt;alert(1)&lt;/script foo="bar"&gt;</code> will not be filtered by 
the function, and <code>alert(1)</code> will be executed by a browser if the string is rendered as HTML.
```
I know the "but" was meant to stand in contrast to "being filtered", but somehow it doesn't read that way. It ends up sounding like what comes after "but" is somehow in contrast to _not_ being filtered.